tree/dataframe/test/dataframe_datasetspec.py

```diff
@@ -0,0 +1,93 @@
+                            .Define("z", [](ULong64_t e) { return e + 100; }, {"rdfentry_"});
+        dfWriter0.Range(0, 2).Snapshot<ULong64_t>("subTree", "PYspecTestFile2.root", {"z"});
+        dfWriter0.Range(2, 4).Snapshot<ULong64_t>("subTree", "PYspecTestFile3.root", {"z"});
+        dfWriter0.Range(4, 5).Snapshot<ULong64_t>("subTree", "PYspecTestFile4.root", {"z"});
+        dfWriter0.Range(0, 2).Snapshot<ULong64_t>("subTree1", "PYspecTestFile5.root", {"z"});
+        dfWriter0.Range(2, 4).Snapshot<ULong64_t>("subTree2", "PYspecTestFile6.root", {"z"});
+        dfWriter0.Snapshot<ULong64_t>("anotherTree", "PYspecTestFile7.root", {"z"});
+        }'''
+        ROOT.gInterpreter.Calc(code)
+
+    def tearDown(self):
+        for i in range(2, 8):
+            ROOT.gSystem.Unlink("PYspecTestFile" + str(i) + ".root")
+
+    # a test involving all 3 constructors, all possible valid ranges, all possible friends
+    def test_General(self):
+        ranges = [REntryRange(), REntryRange(1, 4), REntryRange(2,4), REntryRange(100), REntryRange(1, 100),
+                  REntryRange(2, 2), REntryRange(7, 7)]
+        expectedRess = [[100, 101, 102, 103, 104], [101, 102, 103], [102, 103], [100, 101, 102, 103, 104],
+                        [101, 102, 103, 104], [], []]
+        specs = []
+        
+        for r in ranges:
+            specs.append([])
+            specs[-1].append(RDatasetSpec("anotherTree", "PYspecTestFile7.root", r))
+            specs[-1].append(RDatasetSpec("subTree", ["PYspecTestFile2.root",
+                                                 "PYspecTestFile3.root",
+                                                 "PYspecTestFile4.root"], r))
+            specs[-1].append(RDatasetSpec([("subTree1", "PYspecTestFile5.root"),
+                                      ("subTree2", "PYspecTestFile6.root"),
+                                      ("subTree", "PYspecTestFile4.root")], r))
+
+            # for each spec add all possible combination of friends
+            for s in specs[-1]:
+                s.AddFriend("anotherTree", "PYspecTestFile7.root", "friendTree")
+                s.AddFriend("subTree", ["PYspecTestFile2.root",
+                                        "PYspecTestFile3.root",
+                                        "PYspecTestFile4.root"], "friendChain1")
+                s.AddFriend([("subTree1", "PYspecTestFile5.root"),
+                             ("subTree2", "PYspecTestFile6.root"),
+                             ("subTree", "PYspecTestFile4.root")], "friendChainN")
+
+        for i in range(len(ranges)):
+            for s in specs[i]:
+                rdf = RDataFrame(RDatasetSpec(s))
+
+                if self.legacy_pyroot:
+                    resP = rdf.Take("ULong64_t")("z")
+                    fr1P = rdf.Take("ULong64_t")("friendTree.z")
+                    fr2P = rdf.Take("ULong64_t")("friendChain1.z")
+                    fr3 = rdf.Take("ULong64_t")("friendChainN.z").GetValue()
```

Comment:
Why `fr3` calls `GetValue` in the same line while `fr1` and `fr2` book the operation and then call `GetValue` afterwards?

Reply:
@eguiraud pointed that the alternative triggers the event loop 4 times (as I am asking for new lazy actions, after a trigger). Now I afaik I am going to trigger the evenet loop once, no?

Reply:
Ah yes I understand what you mean. I still find it weird to leave one `GetValue` there and the others without it, I would find it clearer if you were booking everything, then calling `GetValue` separately on everything. But still, the code as it is should be triggering only once the event loop, indeed